packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -145,46 +145,67 @@ abstract class GlobalKey<T extends State<StatefulWidget>> extends Key {
       _registry.remove(this);
   }
 
-  void _debugReserveFor(Element parent) {
+  void _debugReserveFor(Element parent, Element newChild, Element oldChild) {
     assert(() {
       assert(parent != null);
-      if (_debugReservations.containsKey(this) && _debugReservations[this] != parent) {
-        // Reserving a new parent while the old parent is not attached is ok.
-        // This can happen when a renderObject detaches and re-attaches to rendering
-        // tree multiple times.
-        if (_debugReservations[this].renderObject?.attached == false) {
-          _debugReservations[this] = parent;
-          return true;
-        }
-        // It's possible for an element to get built multiple times in one
-        // frame, in which case it'll reserve the same child's key multiple
-        // times. We catch multiple children of one widget having the same key
-        // by verifying that an element never steals elements from itself, so we
-        // don't care to verify that here as well.
-        final String older = _debugReservations[this].toString();
-        final String newer = parent.toString();
-        if (older != newer) {
-          throw FlutterError.fromParts(<DiagnosticsNode>[
-            ErrorSummary('Multiple widgets used the same GlobalKey.'),
-            ErrorDescription(
-              'The key $this was used by multiple widgets. The parents of those widgets were:\n'
-              '- $older\n'
-              '- $newer\n'
-              'A GlobalKey can only be specified on one widget at a time in the widget tree.'
-            ),
-          ]);
-        }
-        throw FlutterError.fromParts(<DiagnosticsNode>[
-          ErrorSummary('Multiple widgets used the same GlobalKey.'),
-          ErrorDescription(
-            'The key $this was used by multiple widgets. The parents of those widgets were '
-            'different widgets that both had the following description:\n'
-            '  $parent\n'
-            'A GlobalKey can only be specified on one widget at a time in the widget tree.'
-          ),
-        ]);
-      }
-      _debugReservations[this] = parent;
+      assert(newChild != null);
+      _debugReservations[parent] ??= <Element, GlobalKey>{};
+      // Removes the old children reservation if exist.
+      _debugReservations[parent].remove(oldChild);
```

Comment:
if early rebuild, the layoutbuilder must update the parent with the new child. In this case, the old child reservation will be removed.

Reply:
I don't quite understand how this gets around the problem of masking a real duplicate key issue. Can you elaborate? If I assign key X to child A of parent P, and then key X to child B of parent P (in the same build phase), how does the framework detect that error?

Reply:
That should be detected in _debugVerifyIllFatedPopulation. I can actually detect that in _debugVerifyGlobalKeyReservation, but i intentionally dropped it because that is not in the scope of this check. see line 178 where this check will drop it ``` _keyToParent[key] != parent```

Reply:
Right but doesn't the line quoted above drop the part of _debugReservations that would have triggered that check?

Maybe the right answer here is to include a test with this PR for this case? Or point me to an existing test that does so already, if there is one, and I can try that test with this PR if I want to see the logic in action.

Reply:
Previously we only throw when we have the same key reserve to different parents. See line 158 in the original file. This line ```_debugReservations[this] != parent``` will drop the case you mentioned.
We do have a test the replicate this case. See https://github.com/flutter/flutter/blob/4938ea03858e178541d723dff430e6c9d044e6ae/packages/flutter/test/widgets/framework_test.dart#L139